### ISSUE TITLE
[Extension Installer] Extension installer is slow and memory heavy

### ISSUE THREAD

**Issue by tkotosz** [phpactor/phpactor#1025]
Extension installer is slow and memory heavy due to the embedded composer. I was playing around with this a couple of weeks ago and turns out embedding composer v2 instead makes it incredibly fast and doesn't suffer from the memory problem either.

BUT (there is always a but) in composer v2 the `setAdditionalInstalledRepository` method no longer exists. As far as I can tell you used this method to avoid installing dependencies in the extensions folder which are already installed in the phpactor root folder (so the dependencies of the main application / build in extensions).

One way to get around it to leave the main application folder completely dependecy-free and just have a wrapper script there which initially installs the application and the default extensions in the extensions folder... that way when installing a non-default extension it will automatically resolve the dependencies correctly since the main app is in the same custom composer json as the extensions.

I am happy to discuss / help with this one, since it is a major change. Btw I played with this here during the easter break: https://github.com/tkotosz/composertest/tree/play-with-composer-2x Basically managed to create a phar file which only contains the embedded composer v2 and instructions for what to install in the custom composer dir, so in your case that would be the phpactor main app and the default extensions basically. - It it not a finalised solution but I am planning to put together a working prototype for a dummy app soon, then I am happy to demo it for you.

**Comment by tkotosz** [comment]
I have played with this a bit more, demo app: https://github.com/tkotosz/fooapp (phar can be downloaded here: https://github.com/tkotosz/fooapp/blob/master/bin/fooapp and should work as it is after `fooapp init`)

The phar basically only responsible for installing the app and its extensions to a unique app folder (folder configured here as input for the phar build process: https://github.com/tkotosz/fooapp/blob/master/cli-app-wrapper-config.json during phar build).

The wrapped app itself which will be installed by the phar executable is this one: https://github.com/tkotosz/fooapp-src

The logic to manage extensions as a wrapper around the provided app is here: https://github.com/tkotosz/cli-app-wrapper and https://github.com/tkotosz/cli-app-wrapper-api .

Basically the app (in this case fooapp) just uses the interfaces provided by the cli-app-wrapper-api to manage extensions (install/remove/list). The interfaces implemented by cli-app-wrapper and compiled (scoped) into the phar along with composer 2.x... seems to be working fine and fast as well. :)

There is some work left todo on it, I just wanted to mention that so far the prototype is promising.

**Comment by dantleech** [comment]
That sounds like a pretty nifty idea. The extension manager is indeed currently a pain and a risk.

The [idea](https://dantleech.com/blog/2018/10/14/rephpactor/) has been to have only a "host" application into which you can install a Phpactor distribution.

I had thought about only doing `composer create-project` but decided against it due to (wrongly or rightly) the user overhead in maintaining their own IDE project.

Having a "shepherd" application to do this sounds perfect.

**Comment by tkotosz** [comment]
Ye I looked into [rephpactor](https://github.com/phpactor/rephpactor) at the beginning, very similar thing indeed. I was trying to generalise it a bit, will see how it goes, I will keep working on it. :)

**Comment by dantleech** [comment]
As phpactor is now a monolith there is no extension manager any more!
